Exception on check command for some plugins, introduced in v3.0.0

I am getting an exception warning (and no expected result) with the "plugman check" command for some plugins since updating from v2.4.1 to v3.0.0. Paper was also updated, from version 1.21.8-28-abf6b6c (MC: 1.21.8), at the same time. None of the affected plugins were updated at the same time.

It seems to always happen for some plugins and never for others. If one of these plugins are present, the "plugman check all" command also fails with a similar exception warning and no normal output.

Steps to reproduce:
1. Add PlugManX and one of the affected plugins
2. Start server
3. Type "plugman check <affected plugin>" or "plugman check all"

What happens:
An exception warning is displayed and no normal result from the command.

What would be expected:
Normal result from the command, no warning.

Other information;
Typing "plugman info <affected plugin>" shows no warning and returns normal command result.

Known affected plugins:
- [Dynmap-EssentialsX](https://mikeprimm.com/dynmap/builds/Dynmap-EssentialsX/) (1.1-SNAPSHOT) _- very old plugin_
- [Essentials](https://essentialsx.net/downloads.html) (2.22.0-dev+18-7c89592) _- recently and frequently updated, commonly used plugin_
- [OpenInv](https://dev.bukkit.org/projects/openinv/files/) (5.1.13) _- recently and frequently updated plugin_
- [Vault](https://www.spigotmc.org/resources/vault.34315/) (1.7.3-b131) _- very very old, but very commonly used, plugin_

Tested plugins that do not seem to be affected:
BlueBridgeCore (2.2), BlueBridgeGP (2.2), BlueMap (5.11), BlueMap-Essentials (1.1.3), BlueMapFloodgate (3.5), BlueMapOfflinePlayerMarkers (3.0), Chunky (1.4.40), ChunkyBorder (1.2.26), CommandTimer (8.13.0), CoreProtect (23.0), DeadChest (4.21.1), DeathCoordinates (1.12.1), DiscordSRV (1.30.0), DragonSlayer (0.14.5_Jeppa), DropHeads (3.10.5.1), dynmap (3.7-SNAPSHOT-991), Dynmap-GriefPrevention (0.90), dynmapHider (1.3), EssentialsChat (2.22.0-dev+18-7c89592), EssentialsGeoIP (2.22.0-dev+18-7c89592), EssentialsSpawn (2.22.0-dev+18-7c89592), floodgate (2.2.4-SNAPSHOT (b118-40d320a)), Geyser-Spigot (2.8.3-SNAPSHOT), GriefPrevention (16.18.4), GSit (2.4.3), InventoryRollbackPlus (1.7.6), LuckPerms (5.5.10), OITogglePersist (5.1.1), PlaceholderAPI (2.11.6), PlugManX (3.0.0), ProtocolLib (5.4.0), TAB (5.2.5), TimeIsMoney (1.9.12), TogglePvp (8.1), ViaVersion (5.5.0-SNAPSHOT)
**NB - some of these give the "Could not fetch plugin data - does the plugin exist as a Spigot Resource?" result, but that is as expected**.

Setup:
```
[bootstrap] Running Java 21 (OpenJDK 64-Bit Server VM 21.0.2+13-LTS; Amazon.com Inc. Corretto-21.0.2.13.1) on Windows Server 2019 10.0 (amd64)
[bootstrap] Loading Paper 1.21.8-39-main@616a879 (2025-08-18T02:25:55Z) for Minecraft 1.21.8
> version
This server is running Paper version 1.21.8-39-main@616a879 (2025-08-18T02:25:55Z) (Implementing API version 1.21.8-R0.1-SNAPSHOT)
You are running the latest version
> version plugmanx
PlugManX version 3.0.0
Plugin manager for Bukkit servers
Authors: rylinaux, and TestAccount666
```

Example output for affected plugin:
```
> plugman check essentials
[09:54:19 INFO]: [PlugMan] Retrieving plugin versions...
[09:54:20 WARN]: [PlugManX] Plugin PlugManX v3.0.0 generated an exception while executing task 7958
java.lang.NullPointerException: null
        at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:904) ~[guava-33.3.1-jre.jar:?]
        at com.google.common.io.Files$FileByteSource.<init>(Files.java:129) ~[guava-33.3.1-jre.jar:?]
        at com.google.common.io.Files$FileByteSource.<init>(Files.java:122) ~[guava-33.3.1-jre.jar:?]
        at com.google.common.io.Files.asByteSource(Files.java:119) ~[guava-33.3.1-jre.jar:?]
        at PlugManX-3.0.0.jar/core.com.rylinaux.plugman.util.updatechecker.CurseForgeUtil.checkUpToDate(CurseForgeUtil.java:124) ~[PlugManX-3.0.0.jar:?]
        at PlugManX-3.0.0.jar/core.com.rylinaux.plugman.util.updatechecker.UpdateUtil.checkUpToDate(UpdateUtil.java:78) ~[PlugManX-3.0.0.jar:?]
        at PlugManX-3.0.0.jar/core.com.rylinaux.plugman.commands.executables.CheckCommand.lambda$execute$3(CheckCommand.java:176) ~[PlugManX-3.0.0.jar:?]
        at org.bukkit.craftbukkit.scheduler.CraftTask.run(CraftTask.java:78) ~[paper-1.21.8.jar:1.21.8-39-616a879]
        at org.bukkit.craftbukkit.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:57) ~[paper-1.21.8.jar:1.21.8-39-616a879]
        at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22) ~[paper-1.21.8.jar:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
```

I tried v3.0.2.-RC.1 and it seems it still has problems with "check" for the same plugins, but it is now a different warning/exception (eg: _Plugin PlugManX v3.0.2-RC.1 generated an exception while executing task 1067
java.lang.RuntimeException: java.io.FileNotFoundException: C:\apps\Minecraft%20Paper%20Sep23\plugins\PlugManX-3.0.2-RC.1.jar (The system cannot find the path specified_):

```
> plugman check all
[12:14:57 INFO]: [PlugMan] Retrieving plugin versions...
[12:14:58 WARN]: [PlugManX] Plugin PlugManX v3.0.2-RC.1 generated an exception while executing task 1043
java.lang.RuntimeException: java.io.FileNotFoundException: C:\apps\Minecraft%20Paper%20Sep23\plugins\PlugManX-3.0.2-RC.1.jar (The system cannot find the path specified)
        at PlugManX-3.0.2-RC.1.jar/core.com.rylinaux.plugman.util.updatechecker.CurseForgeUtil.checkUpToDate(CurseForgeUtil.java:128) ~[PlugManX-3.0.2-RC.1.jar:?]
        at PlugManX-3.0.2-RC.1.jar/core.com.rylinaux.plugman.util.updatechecker.UpdateUtil.checkUpToDate(UpdateUtil.java:77) ~[PlugManX-3.0.2-RC.1.jar:?]
        at PlugManX-3.0.2-RC.1.jar/core.com.rylinaux.plugman.util.updatechecker.UpdateUtil.checkUpToDate(UpdateUtil.java:51) ~[PlugManX-3.0.2-RC.1.jar:?]
        at PlugManX-3.0.2-RC.1.jar/core.com.rylinaux.plugman.commands.executables.CheckCommand.lambda$execute$1(CheckCommand.java:112) ~[PlugManX-3.0.2-RC.1.jar:?]
        at org.bukkit.craftbukkit.scheduler.CraftTask.run(CraftTask.java:78) ~[paper-1.21.8.jar:1.21.8-49-614e9ac]
        at org.bukkit.craftbukkit.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:57) ~[paper-1.21.8.jar:1.21.8-49-614e9ac]
        at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22) ~[paper-1.21.8.jar:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.io.FileNotFoundException: C:\apps\Minecraft%20Paper%20Sep23\plugins\PlugManX-3.0.2-RC.1.jar (The system cannot find the path specified)
        at java.base/java.io.FileInputStream.open0(Native Method) ~[?:?]
        at java.base/java.io.FileInputStream.open(FileInputStream.java:213) ~[?:?]
        at java.base/java.io.FileInputStream.<init>(FileInputStream.java:152) ~[?:?]
        at com.google.common.io.Files$FileByteSource.openStream(Files.java:134) ~[guava-33.3.1-jre.jar:?]
        at com.google.common.io.Files$FileByteSource.openStream(Files.java:122) ~[guava-33.3.1-jre.jar:?]
        at com.google.common.io.ByteSource.copyTo(ByteSource.java:256) ~[guava-33.3.1-jre.jar:?]
        at com.google.common.io.ByteSource.hash(ByteSource.java:340) ~[guava-33.3.1-jre.jar:?]
        at PlugManX-3.0.2-RC.1.jar/core.com.rylinaux.plugman.util.updatechecker.CurseForgeUtil.checkUpToDate(CurseForgeUtil.java:126) ~[PlugManX-3.0.2-RC.1.jar:?]
        ... 9 more
> plugman check chunky
[12:15:00 INFO]: [PlugMan] Retrieving plugin versions...
[12:15:00 INFO]: [PlugMan] Plugin seems to be up-to-date using version "1.4.40".
> plugman check vault
[12:15:04 INFO]: [PlugMan] Retrieving plugin versions...
[12:15:04 WARN]: [PlugManX] Plugin PlugManX v3.0.2-RC.1 generated an exception while executing task 1067
java.lang.RuntimeException: java.io.FileNotFoundException: C:\apps\Minecraft%20Paper%20Sep23\plugins\PlugManX-3.0.2-RC.1.jar (The system cannot find the path specified)
        at PlugManX-3.0.2-RC.1.jar/core.com.rylinaux.plugman.util.updatechecker.CurseForgeUtil.checkUpToDate(CurseForgeUtil.java:128) ~[PlugManX-3.0.2-RC.1.jar:?]
        at PlugManX-3.0.2-RC.1.jar/core.com.rylinaux.plugman.util.updatechecker.UpdateUtil.checkUpToDate(UpdateUtil.java:77) ~[PlugManX-3.0.2-RC.1.jar:?]
        at PlugManX-3.0.2-RC.1.jar/core.com.rylinaux.plugman.commands.executables.CheckCommand.lambda$execute$3(CheckCommand.java:176) ~[PlugManX-3.0.2-RC.1.jar:?]
        at org.bukkit.craftbukkit.scheduler.CraftTask.run(CraftTask.java:78) ~[paper-1.21.8.jar:1.21.8-49-614e9ac]
        at org.bukkit.craftbukkit.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:57) ~[paper-1.21.8.jar:1.21.8-49-614e9ac]
        at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22) ~[paper-1.21.8.jar:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.io.FileNotFoundException: C:\apps\Minecraft%20Paper%20Sep23\plugins\PlugManX-3.0.2-RC.1.jar (The system cannot find the path specified)
        at java.base/java.io.FileInputStream.open0(Native Method) ~[?:?]
        at java.base/java.io.FileInputStream.open(FileInputStream.java:213) ~[?:?]
        at java.base/java.io.FileInputStream.<init>(FileInputStream.java:152) ~[?:?]
        at com.google.common.io.Files$FileByteSource.openStream(Files.java:134) ~[guava-33.3.1-jre.jar:?]
        at com.google.common.io.Files$FileByteSource.openStream(Files.java:122) ~[guava-33.3.1-jre.jar:?]
        at com.google.common.io.ByteSource.copyTo(ByteSource.java:256) ~[guava-33.3.1-jre.jar:?]
        at com.google.common.io.ByteSource.hash(ByteSource.java:340) ~[guava-33.3.1-jre.jar:?]
        at PlugManX-3.0.2-RC.1.jar/core.com.rylinaux.plugman.util.updatechecker.CurseForgeUtil.checkUpToDate(CurseForgeUtil.java:126) ~[PlugManX-3.0.2-RC.1.jar:?]
        ... 8 more
> plugman info plugmanx
[12:15:11 INFO]: [PlugMan] Plugin Information: PlugManX
[12:15:11 INFO]: - Version: 3.0.2-RC.1
[12:15:11 INFO]: - Author(s): rylinaux, TestAccount666
[12:15:11 INFO]: - Status: Enabled
> version
[12:15:13 INFO]: This server is running Paper version 1.21.8-49-main@614e9ac (2025-08-23T21:22:35Z) (Implementing API version 1.21.8-R0.1-SNAPSHOT)
You are running the latest version
Previous version: 1.21.8-39-616a879 (MC: 1.21.8)
```

--- a/plugman-assembly/pom.xml
+++ b/plugman-assembly/pom.xml
@@ -7,7 +7,7 @@
     <parent>
         <groupId>com.rylinaux</groupId>
         <artifactId>PlugManX</artifactId>
-        <version>3.0.1</version>
+        <version>3.0.2</version>
         <relativePath>../pom.xml</relativePath>
     </parent>
 

--- a/plugman-bukkit/pom.xml
+++ b/plugman-bukkit/pom.xml
@@ -7,7 +7,7 @@
     <parent>
         <groupId>com.rylinaux</groupId>
         <artifactId>PlugManX</artifactId>
-        <version>3.0.1</version>
+        <version>3.0.2</version>
         <relativePath>../pom.xml</relativePath>
     </parent>
 

--- a/plugman-bukkit/src/main/java/bukkit/com/rylinaux/plugman/BukkitPlugManInitializer.java
+++ b/plugman-bukkit/src/main/java/bukkit/com/rylinaux/plugman/BukkitPlugManInitializer.java
@@ -30,6 +30,7 @@ import bukkit.com.rylinaux.plugman.auto.BukkitAutoFeatureManager;
 import bukkit.com.rylinaux.plugman.config.BukkitConfigurationProvider;
 import bukkit.com.rylinaux.plugman.config.BukkitPlugManConfigurationManager;
 import bukkit.com.rylinaux.plugman.messaging.BukkitColorFormatter;
+import bukkit.com.rylinaux.plugman.plugin.BukkitPlugin;
 import bukkit.com.rylinaux.plugman.pluginmanager.BukkitPluginManager;
 import bukkit.com.rylinaux.plugman.util.BukkitThreadUtil;
 import core.com.rylinaux.plugman.auto.AutoFeatureManager;
@@ -54,7 +55,7 @@ public class BukkitPlugManInitializer extends BasePlugManInitializer {
     private final PlugManBukkit plugin;
 
     public BukkitPlugManInitializer(PlugManBukkit plugin, ServiceRegistry serviceRegistry, PluginLogger logger) {
-        super(serviceRegistry, logger);
+        super(new BukkitPlugin(plugin), serviceRegistry, logger);
         this.plugin = plugin;
     }
 

--- a/plugman-bukkit/src/main/java/bukkit/com/rylinaux/plugman/pluginmanager/BasePluginManager.java
+++ b/plugman-bukkit/src/main/java/bukkit/com/rylinaux/plugman/pluginmanager/BasePluginManager.java
@@ -29,11 +29,12 @@ package bukkit.com.rylinaux.plugman.pluginmanager;
 import bukkit.com.rylinaux.plugman.PlugManBukkit;
 import bukkit.com.rylinaux.plugman.api.PlugManAPI;
 import core.com.rylinaux.plugman.config.PlugManConfigurationManager;
+import core.com.rylinaux.plugman.plugins.Command;
+import core.com.rylinaux.plugman.plugins.CommandMapWrap;
 import core.com.rylinaux.plugman.plugins.Plugin;
 import core.com.rylinaux.plugman.plugins.PluginManager;
 import core.com.rylinaux.plugman.util.reflection.FieldAccessor;
 import org.bukkit.Bukkit;
-import org.bukkit.command.Command;
 import org.bukkit.command.PluginCommand;
 import org.bukkit.command.SimpleCommandMap;
 import org.bukkit.event.Event;
@@ -130,9 +131,9 @@ public abstract class BasePluginManager implements PluginManager {
      * Common plugin command handling logic.
      */
     protected void handlePluginCommand(Plugin plugin, SimpleCommandMap commandMap,
-                                       Map<String, Command> modifiedKnownCommands,
+                                       CommandMapWrap<org.bukkit.command.Command> modifiedKnownCommands,
                                        Map.Entry<String, Command> entry) {
-        var command = (PluginCommand) entry.getValue();
+        var command = (PluginCommand) entry.getValue().<org.bukkit.command.Command>getHandle();
         var bukkitPlugin = plugin.<org.bukkit.plugin.Plugin>getHandle();
         if (command.getPlugin() != bukkitPlugin) return;
         command.unregister(commandMap);
@@ -143,19 +144,19 @@ public abstract class BasePluginManager implements PluginManager {
      * Common broken command removal logic.
      */
     protected void handleBrokenCommand(Map.Entry<String, Command> entry, SimpleCommandMap commandMap,
-                                       Map<String, Command> modifiedKnownCommands, String loggerName) {
+                                       CommandMapWrap<org.bukkit.command.Command> modifiedKnownCommands, String loggerName) {
         var config = PlugManBukkit.getInstance().<PlugManConfigurationManager>get(PlugManConfigurationManager.class);
 
-        if (config.shouldNotifyOnBrokenCommandRemoval())
-            Logger.getLogger(loggerName).info("Removing broken command '" + entry.getValue().getName() + "'!");
-        entry.getValue().unregister(commandMap);
+        var handle = entry.getValue().<org.bukkit.command.Command>getHandle();
+        if (config.shouldNotifyOnBrokenCommandRemoval()) Logger.getLogger(loggerName).info("Removing broken command '" + handle.getName() + "'!");
+        handle.unregister(commandMap);
         modifiedKnownCommands.remove(entry.getKey());
     }
 
     /**
      * Common command loading logic.
      */
-    protected void scheduleCommandLoading() {
+    protected synchronized void scheduleCommandLoading() {
         Bukkit.getScheduler().runTaskLater(PlugManBukkit.getInstance(), this::syncCommands, 10L);
     }
 
@@ -169,7 +170,7 @@ public abstract class BasePluginManager implements PluginManager {
      * Common data structure for unload operations.
      */
     public record CommonUnloadData(org.bukkit.plugin.PluginManager pluginManager, SimpleCommandMap commandMap, List<org.bukkit.plugin.Plugin> plugins,
-                                   Map<String, org.bukkit.plugin.Plugin> names, Map<String, org.bukkit.command.Command> commands,
+                                   Map<String, org.bukkit.plugin.Plugin> names, CommandMapWrap<org.bukkit.command.Command> commands,
                                    Map<Event, SortedSet<RegisteredListener>> listeners,
                                    boolean reloadListeners) {
     }

--- a/plugman-bukkit/src/main/java/bukkit/com/rylinaux/plugman/pluginmanager/BukkitPluginManager.java
+++ b/plugman-bukkit/src/main/java/bukkit/com/rylinaux/plugman/pluginmanager/BukkitPluginManager.java
@@ -287,7 +287,7 @@ public class BukkitPluginManager extends BasePluginManager {
         var plugins = new ArrayList<String>();
 
         for (var entry : getKnownCommands().entrySet()) {
-            var cl = entry.getValue().getClass().getClassLoader();
+            var cl = entry.getValue().getHandle().getClass().getClassLoader();
 
             if (cl.getClass() != pluginClassLoaderClass) handleNonPluginClassLoaderCommand(entry, command, plugins);
             else handlePluginClassLoaderCommand(entry, command, plugins, cl);

--- a/plugman-bukkit/src/main/java/bukkit/com/rylinaux/plugman/pluginmanager/BukkitPluginManager.java
+++ b/plugman-bukkit/src/main/java/bukkit/com/rylinaux/plugman/pluginmanager/BukkitPluginManager.java
@@ -6,6 +6,7 @@ import bukkit.com.rylinaux.plugman.plugin.BukkitPlugin;
 import core.com.rylinaux.plugman.PluginResult;
 import core.com.rylinaux.plugman.config.PlugManConfigurationManager;
 import core.com.rylinaux.plugman.plugins.Command;
+import core.com.rylinaux.plugman.plugins.CommandMapWrap;
 import core.com.rylinaux.plugman.plugins.Plugin;
 import core.com.rylinaux.plugman.util.StringUtil;
 import core.com.rylinaux.plugman.util.reflection.ClassAccessor;
@@ -196,8 +197,7 @@ public class BukkitPluginManager extends BasePluginManager {
     @Override
     public List<String> getPluginNames(boolean fullName) {
         var plugins = new ArrayList<String>();
-        for (var plugin : Bukkit.getPluginManager().getPlugins())
-            plugins.add(fullName? plugin.getDescription().getFullName() : plugin.getName());
+        for (var plugin : Bukkit.getPluginManager().getPlugins()) plugins.add(fullName? plugin.getDescription().getFullName() : plugin.getName());
         return plugins;
     }
 
@@ -267,10 +267,11 @@ public class BukkitPluginManager extends BasePluginManager {
         var knownCommands = getKnownCommands();
 
         return knownCommands.entrySet().stream()
-                .filter(s -> {
-                    if (s.getKey().contains(":")) return s.getKey().split(":")[0].equalsIgnoreCase(plugin.getName());
+                .filter(entry -> {
+                    var name = entry.getKey();
+                    if (name.contains(":")) return name.split(":")[0].equalsIgnoreCase(plugin.getName());
                     else {
-                        var classLoader = s.getValue().getHandle().getClass().getClassLoader();
+                        var classLoader = entry.getValue().getHandle().getClass().getClassLoader();
                         return classLoader.getClass() == pluginClassLoaderClass && getPluginFromClassLoader.apply(classLoader) == plugin.getHandle();
                     }
                 }).collect(Collectors.toList());
@@ -299,7 +300,8 @@ public class BukkitPluginManager extends BasePluginManager {
     private void handleNonPluginClassLoaderCommand(Map.Entry<String, Command> entry, String command, List<String> plugins) {
         var parts = entry.getKey().split(":");
 
-        if (parts.length == 2 && parts[1].equalsIgnoreCase(command)) Arrays.stream(Bukkit.getPluginManager().getPlugins())
+        if (parts.length != 2 || !parts[1].equalsIgnoreCase(command)) return;
+        Arrays.stream(Bukkit.getPluginManager().getPlugins())
                 .filter(pl -> pl.getName().equalsIgnoreCase(parts[0]))
                 .findFirst().ifPresent(plugin -> plugins.add(plugin.getName()));
     }
@@ -385,42 +387,16 @@ public class BukkitPluginManager extends BasePluginManager {
     }
 
     @Override
-    public Map<String, Command> getKnownCommands() {
+    public CommandMapWrap<org.bukkit.command.Command> getKnownCommands() {
         try {
             var commandMap = getCommandMap();
             var knownCommands = FieldAccessor.<Map<String, org.bukkit.command.Command>>getValue(SimpleCommandMap.class, "knownCommands", commandMap);
 
-            return convertBukkitCommands(knownCommands);
+            return new CommandMapWrap<>(knownCommands, BukkitCommand::new);
         } catch (Exception exception) {
             PlugManBukkit.getInstance().getLogger().log(Level.SEVERE, "Failed to get known commands", exception);
             return null;
         }
-    }
-
-    @Override
-    public void setKnownCommands(Map<String, Command> knownPlugManCommands) {
-        try {
-            var knownCommands = convertPlugManCommands(knownPlugManCommands);
-
-            var commandMap = getCommandMap();
-
-            FieldAccessor.setValue(SimpleCommandMap.class, "knownCommands", commandMap, knownCommands);
-        } catch (Exception exception) {
-            PlugManBukkit.getInstance().getLogger().log(Level.SEVERE, "Failed to set known commands", exception);
-        }
-    }
-
-    @ApiStatus.Internal
-    public Map<String, org.bukkit.command.Command> convertPlugManCommands(Map<String, Command> plugManCommands) {
-        var bukkitCommands = new HashMap<String, org.bukkit.command.Command>();
-        for (var entry : plugManCommands.entrySet()) {
-            var commandName = entry.getKey();
-            var command = entry.getValue();
-
-            bukkitCommands.put(commandName, command.getHandle());
-        }
-
-        return bukkitCommands;
     }
 
     /**
@@ -433,7 +409,7 @@ public class BukkitPluginManager extends BasePluginManager {
     public synchronized PluginResult unload(Plugin plugin) {
         if (!handleGentleUnload(plugin)) return new PluginResult(false, "unload.gentle-failed");
 
-        syncCommands();
+        //syncCommands();
 
         var unloadData = extractPluginManagerData(plugin);
         if (unloadData == null) return new PluginResult(false, "unload.failed");
@@ -468,8 +444,8 @@ public class BukkitPluginManager extends BasePluginManager {
                 reloadlisteners = false;
             }
 
-            var commandMap = FieldAccessor.<SimpleCommandMap>getValue(pluginManager.getClass(), "commandMap", pluginManager);
-            var commands = FieldAccessor.<Map<String, org.bukkit.command.Command>>getValue(SimpleCommandMap.class, "knownCommands", commandMap);
+            var commandMap = getCommandMap();//FieldAccessor.<SimpleCommandMap>getValue(pluginManager.getClass(), "commandMap", pluginManager);
+            var commands = getKnownCommands(); //FieldAccessor.<Map<String, org.bukkit.command.Command>>getValue(SimpleCommandMap.class, "knownCommands", commandMap);
 
             pluginManager.disablePlugin(bukkitPlugin);
 
@@ -486,21 +462,20 @@ public class BukkitPluginManager extends BasePluginManager {
 
         cleanupListeners(plugin, data.listeners(), data.reloadListeners());
         cleanupCommands(plugin, data);
+        syncCommands();
         removeFromPluginLists(plugin, data);
     }
 
     private void cleanupCommands(Plugin plugin, CommonUnloadData data) {
         if (data.commandMap() == null) return;
 
-        var modifiedKnownCommands = new HashMap<>(data.commands());
+        var modifiedKnownCommands = data.commands();
 
-        for (var entry : new HashMap<>(data.commands()).entrySet())
-            if (entry.getValue() instanceof PluginCommand) handlePluginCommand(plugin, data.commandMap(), modifiedKnownCommands, entry);
+        for (var entry : modifiedKnownCommands.asMap().entrySet())
+            if (entry.getValue().getHandle() instanceof PluginCommand) handlePluginCommand(plugin, data.commandMap(), modifiedKnownCommands, entry);
             else handleNonPluginCommand(plugin, data.commandMap(), modifiedKnownCommands, entry);
 
-        var plugManCommands = convertBukkitCommands(modifiedKnownCommands);
-
-        setKnownCommands(plugManCommands);
+        syncCommands();
     }
 
     @ApiStatus.Internal
@@ -518,8 +493,8 @@ public class BukkitPluginManager extends BasePluginManager {
 
     @ApiStatus.Internal
     public void handleNonPluginCommand(Plugin plugin, SimpleCommandMap commandMap,
-                                       Map<String, org.bukkit.command.Command> modifiedKnownCommands,
-                                       Map.Entry<String, org.bukkit.command.Command> entry) {
+                                       CommandMapWrap<org.bukkit.command.Command> modifiedKnownCommands,
+                                       Map.Entry<String, Command> entry) {
         try {
             unregisterNonPluginCommands(plugin, commandMap, modifiedKnownCommands, entry);
         } catch (IllegalStateException exception) {
@@ -529,15 +504,17 @@ public class BukkitPluginManager extends BasePluginManager {
     }
 
     @ApiStatus.Internal
-    public void unregisterNonPluginCommands(Plugin plugin, CommandMap commandMap, Map<String, org.bukkit.command.Command> commands,
-                                            Map.Entry<String, org.bukkit.command.Command> entry) {
+    public void unregisterNonPluginCommands(Plugin plugin, CommandMap commandMap, CommandMapWrap<org.bukkit.command.Command> commands,
+                                            Map.Entry<String, Command> entry) {
         var command = entry.getValue();
-        var pluginField = FieldAccessor.getFirstFieldName(command.getClass(), Plugin.class);
+        var handle = command.<org.bukkit.command.Command>getHandle();
+
+        var pluginField = FieldAccessor.getFirstFieldName(handle.getClass(), Plugin.class);
 
         try {
-            var owningPlugin = FieldAccessor.<org.bukkit.plugin.Plugin>getValue(command.getClass(), pluginField, command);
+            var owningPlugin = FieldAccessor.<org.bukkit.plugin.Plugin>getValue(handle.getClass(), pluginField, handle);
             if (owningPlugin != null && owningPlugin.getName().equalsIgnoreCase(plugin.getName())) {
-                command.unregister(commandMap);
+                handle.unregister(commandMap);
                 commands.remove(entry.getKey());
             }
         } catch (IllegalAccessException exception) {

--- a/plugman-bukkit/src/main/resources/plugin.yml
+++ b/plugman-bukkit/src/main/resources/plugin.yml
@@ -8,6 +8,8 @@ folia-supported: true
 provides: [ PlugMan ]
 libraries:
   - com.fasterxml.jackson.core:jackson-databind:${jackson.version}
+  - com.fasterxml.jackson.core:jackson-core:${jackson.version}
+  - com.fasterxml.jackson.core:jackson-annotations:${jackson.version}
 load: STARTUP
 softdepend:
   - PlugMan

--- a/plugman-bungee/pom.xml
+++ b/plugman-bungee/pom.xml
@@ -7,7 +7,7 @@
     <parent>
         <groupId>com.rylinaux</groupId>
         <artifactId>PlugManX</artifactId>
-        <version>3.0.1</version>
+        <version>3.0.2</version>
         <relativePath>../pom.xml</relativePath>
     </parent>
 

--- a/plugman-bungee/src/main/java/bungee/com/rylinaux/plugman/BungeePlugManInitializer.java
+++ b/plugman-bungee/src/main/java/bungee/com/rylinaux/plugman/BungeePlugManInitializer.java
@@ -30,6 +30,7 @@ import bungee.com.rylinaux.plugman.auto.BungeeAutoFeatureManager;
 import bungee.com.rylinaux.plugman.config.BungeeConfigurationProvider;
 import bungee.com.rylinaux.plugman.config.BungeePlugManConfigurationManager;
 import bungee.com.rylinaux.plugman.messaging.BungeeColorFormatter;
+import bungee.com.rylinaux.plugman.plugin.BungeePlugin;
 import bungee.com.rylinaux.plugman.pluginmanager.BungeePluginManager;
 import bungee.com.rylinaux.plugman.util.BungeeThreadUtil;
 import core.com.rylinaux.plugman.auto.AutoFeatureManager;
@@ -56,7 +57,7 @@ public class BungeePlugManInitializer extends BasePlugManInitializer {
     private final PlugManBungee plugin;
 
     public BungeePlugManInitializer(PlugManBungee plugin, ServiceRegistry serviceRegistry, PluginLogger logger) {
-        super(serviceRegistry, logger);
+        super(new BungeePlugin(plugin), serviceRegistry, logger);
         this.plugin = plugin;
     }
 

--- a/plugman-bungee/src/main/java/bungee/com/rylinaux/plugman/pluginmanager/BungeePluginManager.java
+++ b/plugman-bungee/src/main/java/bungee/com/rylinaux/plugman/pluginmanager/BungeePluginManager.java
@@ -6,7 +6,7 @@ import bungee.com.rylinaux.plugman.plugin.BungeePlugin;
 import com.google.common.collect.Multimap;
 import core.com.rylinaux.plugman.PluginResult;
 import core.com.rylinaux.plugman.config.PlugManConfigurationManager;
-import core.com.rylinaux.plugman.plugins.Command;
+import core.com.rylinaux.plugman.plugins.CommandMapWrap;
 import core.com.rylinaux.plugman.plugins.Plugin;
 import core.com.rylinaux.plugman.plugins.PluginManager;
 import core.com.rylinaux.plugman.util.reflection.FieldAccessor;
@@ -196,55 +196,17 @@ public class BungeePluginManager implements PluginManager {
     }
 
     @Override
-    public Map<String, Command> getKnownCommands() {
+    public CommandMapWrap<net.md_5.bungee.api.plugin.Command> getKnownCommands() {
         try {
             var pluginManager = ProxyServer.getInstance().getPluginManager();
             var commandMap = FieldAccessor.<Map<String, net.md_5.bungee.api.plugin.Command>>getValue(
                     net.md_5.bungee.api.plugin.PluginManager.class, "commandMap", pluginManager);
 
-            if (commandMap == null) return new HashMap<>();
-
-            return convertBungeeCommands(commandMap);
+            return new CommandMapWrap<>(commandMap, BungeeCommand::new);
         } catch (Exception exception) {
             PlugManBungee.getInstance().getLogger().log(Level.SEVERE, "Failed to get known commands", exception);
-            return new HashMap<>();
+            return null;
         }
-    }
-
-    @Override
-    public void setKnownCommands(Map<String, Command> knownCommands) {
-        try {
-            var pluginManager = ProxyServer.getInstance().getPluginManager();
-            var bungeeCommands = convertPlugManCommands(knownCommands);
-
-            FieldAccessor.setValue(net.md_5.bungee.api.plugin.PluginManager.class, "commandMap", pluginManager, bungeeCommands);
-        } catch (Exception exception) {
-            PlugManBungee.getInstance().getLogger().log(Level.SEVERE, "Failed to set known commands", exception);
-        }
-    }
-
-    private Map<String, Command> convertBungeeCommands(Map<String, net.md_5.bungee.api.plugin.Command> bungeeCommands) {
-        var plugManCommands = new HashMap<String, Command>();
-        for (var entry : bungeeCommands.entrySet()) {
-            var commandName = entry.getKey();
-            var command = entry.getValue();
-
-            plugManCommands.put(commandName, new BungeeCommand(command));
-        }
-
-        return plugManCommands;
-    }
-
-    private Map<String, net.md_5.bungee.api.plugin.Command> convertPlugManCommands(Map<String, Command> plugManCommands) {
-        var bungeeCommands = new HashMap<String, net.md_5.bungee.api.plugin.Command>();
-        for (var entry : plugManCommands.entrySet()) {
-            var commandName = entry.getKey();
-            var command = entry.getValue();
-
-            bungeeCommands.put(commandName, command.getHandle());
-        }
-
-        return bungeeCommands;
     }
 
     @Override

--- a/plugman-core/pom.xml
+++ b/plugman-core/pom.xml
@@ -8,7 +8,7 @@
     <parent>
         <groupId>com.rylinaux</groupId>
         <artifactId>PlugManX</artifactId>
-        <version>3.0.1</version>
+        <version>3.0.2</version>
         <relativePath>../pom.xml</relativePath>
     </parent>
 

--- a/plugman-core/src/main/java/core/com/rylinaux/plugman/initialization/BasePlugManInitializer.java
+++ b/plugman-core/src/main/java/core/com/rylinaux/plugman/initialization/BasePlugManInitializer.java
@@ -30,6 +30,7 @@ import core.com.rylinaux.plugman.auto.AutoFeatureManager;
 import core.com.rylinaux.plugman.config.PlugManConfigurationManager;
 import core.com.rylinaux.plugman.file.messaging.MessageFormatter;
 import core.com.rylinaux.plugman.logging.PluginLogger;
+import core.com.rylinaux.plugman.plugins.Plugin;
 import core.com.rylinaux.plugman.plugins.PluginManager;
 import core.com.rylinaux.plugman.services.ServiceRegistry;
 import core.com.rylinaux.plugman.util.ThreadUtil;
@@ -57,9 +58,18 @@ public abstract class BasePlugManInitializer {
     @Getter
     protected final File dataFolder = Path.of("plugins", "PlugManX").toFile();
 
-    protected BasePlugManInitializer(ServiceRegistry serviceRegistry, PluginLogger logger) {
+    protected BasePlugManInitializer(Plugin plugMan, ServiceRegistry serviceRegistry, PluginLogger logger) {
         this.serviceRegistry = serviceRegistry;
         this.logger = logger;
+
+        checkForDevVersion(plugMan);
+    }
+
+    private void checkForDevVersion(Plugin plugin) {
+        var version = plugin.getVersion();
+        if (!version.toLowerCase().contains("dev") && !version.toLowerCase().contains("-rc.")) return;
+
+        logger.warning("You are running a development version (" + version + ") of PlugMan. This is not recommended for production use.");
     }
 
     /**

--- a/plugman-core/src/main/java/core/com/rylinaux/plugman/plugins/CommandMapWrap.java
+++ b/plugman-core/src/main/java/core/com/rylinaux/plugman/plugins/CommandMapWrap.java
@@ -1,0 +1,66 @@
+package core.com.rylinaux.plugman.plugins;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+public class CommandMapWrap<T> {
+    private final Map<String, Command> commands = new HashMap<>();
+    private final Map<String, T> knownCommands;
+
+    public CommandMapWrap(Map<String, T> knownCommands, Function<T, ? extends Command> pluginCommandFactory) {
+        this.knownCommands = knownCommands;
+
+        // Note: Never use `forEach` here. The implementation of `forEach` seems to be a no-op
+        for (var entry : knownCommands.entrySet()) commands.put(entry.getKey(), pluginCommandFactory.apply(entry.getValue()));
+    }
+
+    public Command get(String key) {
+        return commands.get(key);
+    }
+
+    public Command put(String key, Command value) {
+        if (!knownCommands.containsKey(key)) knownCommands.put(key, value.getHandle());
+        return commands.put(key, value);
+    }
+
+    public void putAll(Map<? extends String, ? extends Command> otherMap) {
+        otherMap.forEach((name, command) -> knownCommands.put(name, command.getHandle()));
+        commands.putAll(otherMap);
+    }
+
+    public Command remove(String key) {
+        knownCommands.remove(key);
+        return commands.remove(key);
+    }
+
+    public boolean remove(String key, Command value) {
+        knownCommands.remove(key, value.<T>getHandle());
+        return commands.remove(key, value);
+    }
+
+    public boolean containsKey(String key) {
+        return commands.containsKey(key);
+    }
+
+    public int size() {
+        return commands.size();
+    }
+
+    public boolean isEmpty() {
+        return commands.isEmpty();
+    }
+
+    public Map<String, Command> asMap() {
+        return Map.copyOf(commands);
+    }
+
+    public Set<Map.Entry<String, Command>> entrySet() {
+        return commands.entrySet();
+    }
+
+    public Set<String> keySet() {
+        return commands.keySet();
+    }
+}

--- a/plugman-core/src/main/java/core/com/rylinaux/plugman/plugins/PluginManager.java
+++ b/plugman-core/src/main/java/core/com/rylinaux/plugman/plugins/PluginManager.java
@@ -29,7 +29,6 @@ package core.com.rylinaux.plugman.plugins;
 import core.com.rylinaux.plugman.PluginResult;
 
 import java.util.List;
-import java.util.Map;
 import java.util.Set;
 
 public interface PluginManager {
@@ -162,16 +161,7 @@ public interface PluginManager {
      */
     PluginResult load(String name);
 
-    Map<String, Command> getKnownCommands();
-
-    /**
-     * Changes the knownCommands field in CommandMap
-     *
-     * @param knownCommands the modified known commands
-     * @throws RuntimeException if command map or known commands field cannot be found
-     */
-    void setKnownCommands(Map<String, Command> knownCommands);
-
+    CommandMapWrap<?> getKnownCommands();
 
     /**
      * Loads and enables a plugin.

--- a/plugman-core/src/main/java/core/com/rylinaux/plugman/util/updatechecker/CurseForgeUtil.java
+++ b/plugman-core/src/main/java/core/com/rylinaux/plugman/util/updatechecker/CurseForgeUtil.java
@@ -40,9 +40,11 @@ import lombok.experimental.UtilityClass;
 import java.io.File;
 import java.io.IOException;
 import java.net.URI;
+import java.net.URLDecoder;
 import java.net.http.HttpClient;
 import java.net.http.HttpRequest;
 import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
 import java.time.Duration;
 import java.util.Map;
 import java.util.TreeMap;
@@ -121,7 +123,13 @@ public class CurseForgeUtil {
 
         var file = plugin.getFile();
         // Fetch file manually, not sure why `Plugin#getFile` can fail in the first place, tbh
-        if (file == null) file = new File(plugin.getClass().getProtectionDomain().getCodeSource().getLocation().getFile());
+        if (file == null) {
+            var path = plugin.getClass()
+                    .getProtectionDomain().getCodeSource()
+                    .getLocation().getFile();
+            path = URLDecoder.decode(path, StandardCharsets.UTF_8);
+            file = new File(path);
+        }
         try {
             currentPluginHashCode = Files.asByteSource(file).hash(Hashing.md5());
         } catch (IOException e) {

--- a/plugman-paper/pom.xml
+++ b/plugman-paper/pom.xml
@@ -7,7 +7,7 @@
     <parent>
         <groupId>com.rylinaux</groupId>
         <artifactId>PlugManX</artifactId>
-        <version>3.0.1</version>
+        <version>3.0.2</version>
         <relativePath>../pom.xml</relativePath>
     </parent>
 

--- a/plugman-paper/src/main/java/paper/com/rylinaux/plugman/commands/PaperCommandCreator.java
+++ b/plugman-paper/src/main/java/paper/com/rylinaux/plugman/commands/PaperCommandCreator.java
@@ -1,7 +1,6 @@
 package paper.com.rylinaux.plugman.commands;
 
 import bukkit.com.rylinaux.plugman.PlugManBukkit;
-import bukkit.com.rylinaux.plugman.commands.CommandCreator;
 import io.papermc.paper.command.brigadier.BasicCommand;
 import io.papermc.paper.command.brigadier.CommandSourceStack;
 import org.bukkit.command.CommandExecutor;
@@ -11,12 +10,21 @@ import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
 
-public class PaperCommandCreator extends CommandCreator {
+public class PaperCommandCreator extends OldPaperCommandCreator {
 
     @Override
     public void registerCommand(String commandName, CommandExecutor executor, TabCompleter tabCompleter, String... aliases) {
-        PlugManBukkit.getInstance().registerCommand(commandName, createCommand(executor, tabCompleter, commandName));
-        Arrays.stream(aliases).forEach(alias -> PlugManBukkit.getInstance().registerCommand(alias, createCommand(executor, tabCompleter, alias)));
+        try {
+            PlugManBukkit.getInstance().registerCommand(commandName, createCommand(executor, tabCompleter, commandName));
+            Arrays.stream(aliases).forEach(alias -> PlugManBukkit.getInstance().registerCommand(alias, createCommand(executor, tabCompleter, alias)));
+        } catch (Throwable throwable) {
+            if (!(throwable instanceof ClassNotFoundException
+                    || throwable instanceof NoClassDefFoundError
+                    || throwable instanceof NoSuchMethodError
+                    || throwable instanceof NoSuchMethodException)) throw new RuntimeException(throwable);
+
+            super.registerCommand(commandName, executor, tabCompleter, aliases);
+        }
     }
 
     private BasicCommand createCommand(CommandExecutor executor, TabCompleter tabCompleter, String label) {

--- a/plugman-paper/src/main/java/paper/com/rylinaux/plugman/pluginmanager/PaperPluginManager.java
+++ b/plugman-paper/src/main/java/paper/com/rylinaux/plugman/pluginmanager/PaperPluginManager.java
@@ -39,13 +39,13 @@ import core.com.rylinaux.plugman.util.reflection.MethodAccessor;
 import core.com.rylinaux.plugman.util.tuples.Tuple;
 import io.papermc.paper.plugin.configuration.PluginMeta;
 import lombok.experimental.Delegate;
+import org.bukkit.command.Command;
 import org.bukkit.command.PluginCommand;
 import org.yaml.snakeyaml.error.YAMLException;
 
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Path;
-import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
@@ -158,7 +158,7 @@ public class PaperPluginManager extends BasePluginManager {
 
         var target = loadPluginWithPaper(pluginFile);
         if (target == null) {
-            target = loadAndEnablePlugin(pluginFile);
+            target = _bukkitPluginManager.loadAndEnablePlugin(pluginFile);
             if (target == null) return new PluginResult(false, "load.invalid-plugin");
         }
 
@@ -202,7 +202,7 @@ public class PaperPluginManager extends BasePluginManager {
 
 
     @Override
-    protected void scheduleCommandLoading() {
+    protected synchronized void scheduleCommandLoading() {
         if (isFolia()) {
             var foliaLib = new com.tcoded.folialib.FoliaLib(PlugManBukkit.getInstance());
             foliaLib.getScheduler().runLater(this::syncCommands, 500, TimeUnit.MILLISECONDS);
@@ -231,7 +231,7 @@ public class PaperPluginManager extends BasePluginManager {
     public Tuple<CommonUnloadData, PluginResult> unloadWithPaper(Plugin plugin) {
         if (!handleGentleUnload(plugin)) return new Tuple<>(null, new PluginResult(false, "unload.gentle-failed"));
 
-        var unloadData = setupUnloadData(plugin);
+        var unloadData = extractPluginManagerData(plugin);
         if (unloadData == null) return new Tuple<>(null, new PluginResult(false, "unload.failed"));
 
         cleanupListeners(plugin, unloadData);
@@ -239,12 +239,6 @@ public class PaperPluginManager extends BasePluginManager {
         removeFromPluginLists(plugin, unloadData);
 
         return new Tuple<>(unloadData, new PluginResult(true, "unload.common-success"));
-    }
-
-
-    private CommonUnloadData setupUnloadData(Plugin plugin) {
-        syncCommands();
-        return _bukkitPluginManager.extractPluginManagerData(plugin);
     }
 
     private void cleanupListeners(Plugin plugin, CommonUnloadData data) {
@@ -255,10 +249,12 @@ public class PaperPluginManager extends BasePluginManager {
     private void cleanupCommands(Plugin plugin, CommonUnloadData data) {
         if (data.commandMap() == null) return;
 
-        var modifiedKnownCommands = new HashMap<>(data.commands());
+        var modifiedKnownCommands = data.commands();
 
-        for (var entry : new HashMap<>(data.commands()).entrySet()) {
-            if (entry.getValue() instanceof PluginCommand command) {
+        for (var entry : modifiedKnownCommands.asMap().entrySet()) {
+            var handle = entry.getValue().<Command>getHandle();
+
+            if (handle instanceof PluginCommand command) {
                 if (command.getPlugin() == plugin.getHandle()) {
                     command.unregister(data.commandMap());
                     modifiedKnownCommands.remove(entry.getKey());
@@ -267,20 +263,19 @@ public class PaperPluginManager extends BasePluginManager {
             }
 
             try {
-                unregisterNonPluginCommands(plugin, data.commandMap(), modifiedKnownCommands, entry);
+                _bukkitPluginManager.unregisterNonPluginCommands(plugin, data.commandMap(), modifiedKnownCommands, entry);
             } catch (IllegalStateException exception) {
                 if (exception.getMessage().equalsIgnoreCase("zip file closed")) {
                     var config = PlugManBukkit.getInstance().<PlugManConfigurationManager>get(PlugManConfigurationManager.class);
                     if (config.shouldNotifyOnBrokenCommandRemoval())
-                        Logger.getLogger(PaperPluginManager.class.getName()).info("Removing broken command '" + entry.getValue().getName() + "'!");
-                    entry.getValue().unregister(data.commandMap());
+                        Logger.getLogger(PaperPluginManager.class.getName()).info("Removing broken command '" + handle.getName() + "'!");
+                    handle.unregister(data.commandMap());
                     modifiedKnownCommands.remove(entry.getKey());
                 }
             }
         }
 
-        var plugManCommands = convertBukkitCommands(modifiedKnownCommands);
-        setKnownCommands(plugManCommands);
+        syncCommands();
     }
 
     private void cleanupPaperPluginManager(Plugin plugin) {

--- a/plugman-velocity/pom.xml
+++ b/plugman-velocity/pom.xml
@@ -7,7 +7,7 @@
     <parent>
         <groupId>com.rylinaux</groupId>
         <artifactId>PlugManX</artifactId>
-        <version>3.0.1</version>
+        <version>3.0.2</version>
         <relativePath>../pom.xml</relativePath>
     </parent>
 

--- a/plugman-velocity/src/main/java/velocity/com/rylinaux/plugman/PlugManVelocity.java
+++ b/plugman-velocity/src/main/java/velocity/com/rylinaux/plugman/PlugManVelocity.java
@@ -52,7 +52,7 @@ import java.nio.file.Path;
 @Plugin(
         id = "plugmanvelocity",
         name = "PlugManVelocity",
-        version = "3.0.2-RC.1",
+        version = "3.0.2",
         description = "Plugin manager for Velocity servers.",
         authors = {"rylinaux", "TestAccount666"}
 )

--- a/plugman-velocity/src/main/java/velocity/com/rylinaux/plugman/PlugManVelocity.java
+++ b/plugman-velocity/src/main/java/velocity/com/rylinaux/plugman/PlugManVelocity.java
@@ -33,6 +33,7 @@ import com.velocitypowered.api.event.proxy.ProxyShutdownEvent;
 import com.velocitypowered.api.plugin.Plugin;
 import com.velocitypowered.api.plugin.annotation.DataDirectory;
 import com.velocitypowered.api.proxy.ProxyServer;
+import com.velocitypowered.api.plugin.PluginContainer;
 import core.com.rylinaux.plugman.services.ServiceRegistry;
 import lombok.Getter;
 import lombok.experimental.Delegate;
@@ -49,11 +50,11 @@ import java.nio.file.Path;
  * @author rylinaux
  */
 @Plugin(
-    id = "plugmanvelocity",
-    name = "PlugManVelocity",
-    version = "3.0.1",
-    description = "Plugin manager for Velocity servers.",
-    authors = {"rylinaux", "TestAccount666"}
+        id = "plugmanvelocity",
+        name = "PlugManVelocity",
+        version = "3.0.2-RC.1",
+        description = "Plugin manager for Velocity servers.",
+        authors = {"rylinaux", "TestAccount666"}
 )
 public final class PlugManVelocity {
 
@@ -64,12 +65,16 @@ public final class PlugManVelocity {
     @Delegate
     private ServiceRegistry serviceRegistry;
 
+    private final PluginContainer container;
     private final ProxyServer server;
     private final Logger logger;
     private final Path dataDirectory;
 
+    private VelocityPlugManInitializer initializer;
+
     @Inject
-    public PlugManVelocity(ProxyServer server, Logger logger, @DataDirectory Path dataDirectory) {
+    public PlugManVelocity(PluginContainer container, ProxyServer server, Logger logger, @DataDirectory Path dataDirectory) {
+        this.container = container
         this.server = server;
         this.logger = logger;
         this.dataDirectory = dataDirectory;
@@ -81,7 +86,7 @@ public final class PlugManVelocity {
 
         serviceRegistry = new ServiceRegistry();
         var pluginLogger = new VelocityPluginLogger(logger);
-        var initializer = new VelocityPlugManInitializer(this, serviceRegistry, pluginLogger);
+        initializer = new VelocityPlugManInitializer(this, container, serviceRegistry, pluginLogger);
 
         initializer.initializeCoreServices();
         initializer.setupMessaging();
@@ -95,7 +100,6 @@ public final class PlugManVelocity {
     public void onProxyShutdown(ProxyShutdownEvent event) {
         instance = null;
         var pluginLogger = new VelocityPluginLogger(logger);
-        var initializer = new VelocityPlugManInitializer(this, serviceRegistry, pluginLogger);
         initializer.cleanup();
         server.getCommandManager().unregister("plugman");
     }

--- a/plugman-velocity/src/main/java/velocity/com/rylinaux/plugman/VelocityPlugManInitializer.java
+++ b/plugman-velocity/src/main/java/velocity/com/rylinaux/plugman/VelocityPlugManInitializer.java
@@ -40,6 +40,8 @@ import velocity.com.rylinaux.plugman.config.VelocityPlugManConfigurationManager;
 import velocity.com.rylinaux.plugman.messaging.VelocityColorFormatter;
 import velocity.com.rylinaux.plugman.pluginmanager.VelocityPluginManager;
 import velocity.com.rylinaux.plugman.util.VelocityThreadUtil;
+import velocity.com.rylinaux.plugman.plugin.VelocityPlugin;
+import com.velocitypowered.api.plugin.PluginContainer;
 
 import java.io.File;
 import java.io.IOException;
@@ -54,8 +56,8 @@ public class VelocityPlugManInitializer extends BasePlugManInitializer {
 
     private final PlugManVelocity plugin;
 
-    public VelocityPlugManInitializer(PlugManVelocity plugin, ServiceRegistry serviceRegistry, PluginLogger logger) {
-        super(serviceRegistry, logger);
+    public VelocityPlugManInitializer(PlugManVelocity plugin, PluginContainer container, ServiceRegistry serviceRegistry, PluginLogger logger) {
+        super(new VelocityPlugin(container, plugin), serviceRegistry, logger);
         this.plugin = plugin;
     }
 

--- a/pom.xml
+++ b/pom.xml
@@ -6,7 +6,7 @@
 
     <groupId>com.rylinaux</groupId>
     <artifactId>PlugManX</artifactId>
-    <version>3.0.1</version>
+    <version>3.0.2</version>
     <packaging>pom</packaging>
 
     <modules>
@@ -57,7 +57,7 @@
 
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-        <plugman.version>3.0.1</plugman.version>
+        <plugman.version>3.0.2</plugman.version>
         <manifold.version>2025.1.26</manifold.version>
         <jackson.version>2.13.5</jackson.version>
     </properties>

